documentation/ArrayExpressionLoweringToFIR.md

```diff
@@ -0,0 +1,315 @@
+the structure described above.
+
+TODO: discuss and justify this structure
+### Proposed level of sub-expression inlining
+Several alternatives regarding "inlining" of sub-expression computation in their parent expression `fir.loop`
+are discussed below. The plan is to start by the easiest and to see if FIR optimizers cannot re-group
+the sub-expression into one loop when it is possible. 
+
+#### One Array Expression Node - One Loop
+Every array node of an expression tree is lowered to its own loop
+and a temp storage created for it.
+That is `x + y + z` where `x`, `y` and `z` are arrays ends up as two loops, the first computing
+`x+z` and storing this in a temp array `tmp1`, the second loop computes `x + tmp1` into a second temp `tmp2`.
+
+Pros: Very easy lowering. We do not need to keep of the current `fir.loop` we should emit code into(it must be the one we just created for the node we are visiting)
+, and assuming we are going through the expression tree in a postfix way, we do not need to keep track of where to emit sub-expressions, they already were.
+Cons: Will generate many loops and temporaries that we know without deep analysis are not needed.
+
+#### One "Perfectly" Elemental Expression Sub-Tree - One Loop
+`x + y + z` is lowered into a single loop that directly computes `x(i) + y(i) + z(i)` for each element and store it into a single temp.
+When one of the array sub-expression is not elemental, it is computed on its own outside of the loop and stored into another temp that is then accessed before the loop.
+For instance, `x + foo(y) + z` where `foo(y)` is an array ends up:
+- compute `foo(y)` into `tmp1`
+- compute `x + tmp1 + z`
+With this alternative, transformational such as `transpose` are still treated as `foo`.
+Pros: Generates less FIR, the original expression structure is better preserved inside the loop.
+Cons: Slightly more complex lowering, we need to keep track of the loop where we want to emit the "inlined" operations and where to insert
+the computation of sub-expressions that are not "inlined".
+
+#### One "Affine" Elemental Expression Sub-Tree - One Loop
+Same as before, but also apply transformationals as index re-mapping instead of computing it outside of the loop.
+For instance `x + transpose(y) + z` becomes a loop computing each `x(i, j) + y(j, i) + z(i, j)`.
+Pros: Generates minimal FIR. Not much more complex that the previous one (mostly the same thing).
+Cons: Requires more knowledge about transformationals, an index mapping state must be propagated while lowering.
+
+
+## Examples
+Below are illustrations of Fortran to FIR lowering based on the above design (with one node, one loop). 
+
+### Simple examples
+Fortran:
+```
+  real :: x(4), y(4)
+  y = x
+```
+Here, `x` expression should not be split into a loop.
+
+FIR:
+```
+ %1 = fir.load %x // that is the whole expression evaluation
+ // This is actually the assignment lowering
+ %fir.store %1 to %y
+```
```

Comment:
It should be pointed out that a direct naive lowering to LLVM here is linear in the size of the array.  So such a lowering would not be ideal for large sized arrays.

Reply:
Which is a long way of saying we probably want to put loads and stores in the loop body.

Reply:
Yes, but skipping this extra loop copying to the left-hand side in assignments require some aliasing analysis between the assignee and expression symbols that I am not sure belong to the expression lowering (it is an open question, maybe some small analysis can already be done).
Typically in `x = transpose(x)`, we cannot directly store inside `x`, hence the copy by default to be safe for now.

Reply:
Sure, but if `transpose` is a logical access map, there are no stores involved. `x` is simply given a new access map.